examples/grammar_constraint.py

```diff
@@ -57,25 +58,33 @@ async def step(self):
         # Get valid tokens for next step
         valid_token_ids = self.csd.get_valid_tokens()
 
+        # If generation is a complete derivation, allow the end-of-string token
+        if self.csd.is_complete():
+            valid_token_ids += [self.eos_token_id]
+
+        # If no valid next tokens, reject and terminate
+        if len(valid_token_ids) == 0:
+            self.condition(False)
+            return
```

Comment:
@gabegrand This makes sense, but just to check my understanding, we would now expect to never hit this branch -- is that right? Or is there still some case where we can end up with no valid next tokens?

Reply:
Theoretically, yes. I left this check because there are still some corner cases that might arise in practice: with a degenerate grammar, a restricted LLM vocab, or with the wrong start rule it would be possible to encounter a situation where there are no valid next tokens. I figured it’s best to have  coverage for these corner cases.